geomesa-fs/geomesa-fs-tools/src/main/scala/org/locationtech/geomesa/fs/tools/ingest/FsManageMetadataCommand.scala

```diff
@@ -82,6 +97,229 @@ object FsManageMetadataCommand {
+            count += 1
+          }
+          partitions += PartitionMetadata(partition, update.result, None, 0L)
+        }
+        (count, partitions.result)
+      }
+
+      private class TopLevelListWorker(phaser: Phaser, list: RemoteIterator[FileStatus]) extends Runnable {
+        override def run(): Unit = {
+          try {
+            var i = phaser.getRegisteredParties + 1
+            while (list.hasNext && i < PhaserUtils.MaxParties) {
+              val status = list.next
+              val path = status.getPath
+              val name = path.getName
+              if (status.isDirectory) {
+                if (name != FileBasedMetadataFactory.MetadataDirectory &&
+                    partitions.forall(_.exists(p => p == name || p.startsWith(s"$name/")))) {
+                  i += 1
+                  // use a tiered phaser on each directory avoid the limit of 65535 registered parties
+                  pool.submit(new ListWorker(new Phaser(phaser, 1), name, storage.context.fc.listStatus(path)))
+                }
+              } else if (name != MetadataJson.MetadataPath) {
+                onDisk.computeIfAbsent("", computeFunction).put(name, java.lang.Boolean.TRUE)
+              }
+            }
+            if (list.hasNext) {
+              pool.submit(new TopLevelListWorker(new Phaser(phaser, 1), list))
+            }
+          } finally {
+            phaser.arriveAndDeregister()
+          }
+        }
+      }
+
+      private class ListWorker(phaser: Phaser, partition: String, listDirectory: => RemoteIterator[FileStatus])
+          extends Runnable {
+        override def run(): Unit = {
+          try {
+            var i = phaser.getRegisteredParties + 1
+            val iter = listDirectory
+            while (iter.hasNext && i < PhaserUtils.MaxParties) {
+              val status = iter.next
+              val path = status.getPath
+              val name = path.getName
+              if (status.isDirectory) {
+                val nextPartition = s"$partition/$name"
+                if (partitions.forall(_.exists(p => p == nextPartition || p.startsWith(s"$nextPartition/")))) {
+                  i += 1
+                  // use a tiered phaser on each directory avoid the limit of 65535 registered parties
+                  pool.submit(new ListWorker(new Phaser(phaser, 1), nextPartition, storage.context.fc.listStatus(path)))
+                }
+              } else {
+                val leafPartition =
+                  if (storage.metadata.leafStorage) { s"$partition/${StorageUtils.leaf(name)}" } else { partition }
+                if (partitions.forall(_.contains(leafPartition))) {
+                  onDisk.computeIfAbsent(leafPartition, computeFunction).put(name, java.lang.Boolean.TRUE)
+                }
+              }
+            }
+            if (iter.hasNext) {
+              pool.submit(new ListWorker(new Phaser(phaser, 1), partition, iter))
```

Comment:
How should I be thinking about Phasers?  Is the "1" passed into the constructor a count representing the calling thread?

Reply:
that's the number of parties you want to pre-register with the phaser. it's equivalent to `val p = new Phaser(); p.register()`